### PR TITLE
Refactor/exceptions

### PR DESCRIPTION
**Pull request by MrHug**
Removed many (almost all, only the `oauth` stuff remains untouched) of the `throws Exception` that you seem quite fond of :wink: `catch(Exception e)` will have to wait for another time though...

Depends on #437

### CODE REVIEW

File: syncany-lib/src/main/java/org/syncany/operations/ActionFileHandler.java [hunk 1]
```diff
@@ -94,25 +95,27 @@ private void deleteActionFile(ActionRemoteFile actionFile) throws StorageExcepti
 		transferManager.delete(actionFile);
 	}
 
-	private void uploadActionFile(ActionRemoteFile actionFile) throws Exception {
+	private void uploadActionFile(ActionRemoteFile actionFile) throws StorageException, IOException {
 		logger.log(Level.INFO, "Uploading action file: " + actionFile);
 
-		File tempActionFile = File.createTempFile("syncany-action-", ".tmp");
+		File tempActionFile = null;
+		tempActionFile = File.createTempFile("syncany-action-", ".tmp");
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/ActionFileHandler.java]: What is this for?

File: syncany-lib/src/main/java/org/syncany/operations/OperationException.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.syncany.operations;
+
+
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/OperationException.java]: remove \n

File: syncany-lib/src/main/java/org/syncany/operations/OperationException.java [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Wrapper for exceptions that can be thrown when executing operations.
+ * 
+ * @author Stefan Hugtenburg
+ */
+public class OperationException extends Exception {
+
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/OperationException.java]: remove \n

File: syncany-lib/src/main/java/org/syncany/operations/OperationException.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+	private static final long serialVersionUID = -4277673078529131874L;
+
+	/**
+	 * @param e Exception that caused it.
+	 */
+	public OperationException(Exception e) {
+		super(e);
+	}
+
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/OperationException.java]: remove \n

File: syncany-lib/src/main/java/org/syncany/operations/daemon/DaemonOperation.java [hunk 5]
```diff
@@ -70,82 +73,96 @@
...
-		
+
 		switch (options.getAction()) {
 		case LIST:
 			return executeList();
 
 		case ADD:
-			return executeAdd();
-
+			try {
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/daemon/DaemonOperation.java]: The try/catch blocks in the switch/case look odd. Can we maybe just throw OperationException in the execute*() methods?

File: syncany-lib/src/main/java/org/syncany/operations/daemon/handlers/HTTPExchangeException.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+ */
+package org.syncany.operations.daemon.handlers;
+
+/**
+ * Wrapper for vague Exceptions of the HTTP library.
+ * 
+ * @author Stefan Hugtenburg
+ *
+ */
+public class HTTPExchangeException extends Exception {
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/daemon/handlers/HTTPExchangeException.java]: - Rename to `HttpExchangeException`
- Remove \n

File: syncany-lib/src/main/java/org/syncany/operations/daemon/messages/api/XmlMessageFactory.java [hunk 7]
```diff
@@ -61,86 +65,112 @@
...
 		String messageType = getMessageType(messageStr);
 		Class<? extends Message> messageClass = getMessageClass(messageType);
 
-		Message message = serializer.read(messageClass, messageStr);
+		Message message;
+		try {
+			message = serializer.read(messageClass, messageStr);
+		}
+		catch (Exception e) {
+			throw new DeserializableException(e);
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/daemon/messages/api/XmlMessageFactory.java]: - Is there a reason for the two exceptions `DeserializableException` and `SerializableException`? Why not just one?
- Shouldn't it be `SerializationException` instead?

File: syncany-lib/src/main/java/org/syncany/operations/daemon/messages/api/XmlMessageFactory.java [hunk 8]
```diff
@@ -61,86 +65,112 @@
...
 	private static class FileHistoryIdConverter implements Converter<PartialFileHistory.FileHistoryId> {
 		@Override
-		public PartialFileHistory.FileHistoryId read(InputNode node) throws Exception {
-			return PartialFileHistory.FileHistoryId.parseFileId(node.getValue());
+		public PartialFileHistory.FileHistoryId read(InputNode node) throws InvalidXMLNodeException {
+			try {
+				return PartialFileHistory.FileHistoryId.parseFileId(node.getValue());
+			}
+			catch (Exception e) {
+				throw new InvalidXMLNodeException(e);
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/daemon/messages/api/XmlMessageFactory.java]: - Change to `InvalidXmlNodeException`

File: syncany-lib/src/main/java/org/syncany/operations/down/ApplyChangesOperation.java [hunk 9]
```diff
@@ -96,15 +99,25 @@ public OperationResult execute() throws Exception {
 		}
 
 		Set<MultiChunkId> unknownMultiChunks = determineRequiredMultiChunks(actions, winnersDatabase);
-		
-		downloader.downloadAndDecryptMultiChunks(unknownMultiChunks);
+
+		try {
+			downloader.downloadAndDecryptMultiChunks(unknownMultiChunks);
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/down/ApplyChangesOperation.java]: I feel it's more readable if they are in one try/catch block. 
Or they could be in individual methods.

File: syncany-lib/src/main/java/org/syncany/operations/down/DownOperation.java [hunk 10]
```diff
@@ -127,82 +128,98 @@ public DownOperation(Config config, DownOperationOptions options) {
...
-			localDatabase.deleteAll();
+				// Normally, we wouldn't want to commit in the middle of an operation, but unfortunately
+				// we have to, since not committing causes hanging in database operations, since UNCOMMITTED_READ
+				// does not do enough magic to proceed. The commit in itself is not a problem, since we need
+				// to redownload all remote data anyway.
+				try {
+					localDatabase.commit();
+				}
+				catch (SQLException e) {
+					throw new StorageException("Commit failed", e);
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/down/DownOperation.java]: I do not feel confident to comment on the introduction of new exceptions in `down` right now, but if @pimotte has OK'd this I'm fine with it. I'd love some Clean-Codifying of this method/class though if you have time. Probably in another pull though.

File: syncany-lib/src/main/java/org/syncany/operations/down/DownOperation.java [hunk 11]
```diff
@@ -127,82 +128,98 @@ public DownOperation(Config config, DownOperationOptions options) {
...
-					preDeleteFileHistoriesWithLastVersion);
+				persistMuddyMultiChunks(winnersBranch, allBranches, databaseVersionLocations);
+				removeNonMuddyMultiChunks();
 
-			persistMuddyMultiChunks(winnersBranch, allBranches, databaseVersionLocations);
-			removeNonMuddyMultiChunks();
+				localDatabase.writeKnownRemoteDatabases(newRemoteDatabases);
+				localDatabase.commit();
+			}
+			catch (StorageException | SQLException | OperationException e) {
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/down/DownOperation.java]: Same as above, if @pimotte OK'd this, I'm okay to.

File: syncany-lib/src/main/java/org/syncany/operations/down/actions/ChangeFileSystemAction.java [hunk 12]
```diff
@@ -28,78 +29,86 @@
...
 		if (fromFileMatches && !toFileMatches) { // Normal case	
 			// Original file matches, so we can delete it
 			// Create conflict file for winning file, if it exists
 			if (!toFileExists) {
-				logger.log(Level.INFO, "     - (1) Original file matches, target file does NOT match: deleting original file, creating target file at: "+fileVersion2);
-				
-				deleteFile(fileVersion1);	
-				createFileFolderOrSymlink(fileVersion2);						
+				logger.log(Level.INFO,
+						"     - (1) Original file matches, target file does NOT match: deleting original file, creating target file at: "
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/down/actions/ChangeFileSystemAction.java]: Oh my, indention uglyness -- presumably created by the auto formatting?!

File: syncany-lib/src/main/java/org/syncany/operations/down/actions/ChecksumMismatchException.java [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.syncany.operations.down.actions;
+
+/**
+ * Used when checksums of two files do not match.
+ * 
+ * @author Stefan Hugtenburg
+ *
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/down/actions/ChecksumMismatchException.java]: Removing \n's

File: syncany-lib/src/main/java/org/syncany/operations/down/actions/FileCreatingFileSystemAction.java [hunk 14]
```diff
@@ -70,20 +72,26 @@ protected void createFolder(FileVersion targetFileVersion) throws Exception {
 		}
 	}
 
-	protected void createFile(FileVersion reconstructedFileVersion) throws Exception {
-		File reconstructedFileInCache = assembleFileToCache(reconstructedFileVersion);		
-		moveFileToFinalLocation(reconstructedFileInCache, reconstructedFileVersion);	
+	protected void createFile(FileVersion reconstructedFileVersion) throws IOException {
+		File reconstructedFileInCache;
+		try {
+			reconstructedFileInCache = assembleFileToCache(reconstructedFileVersion);
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/down/actions/FileCreatingFileSystemAction.java]: I generally dislike this pattern: 

```
X var;
try {
   X = ...;
}
catch (..) {
}
```

Couldn't we just move the `moveFileToFinalLocation` inside the `try`?

Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/down/actions/FileCreatingFileSystemAction.java]: I'm also a fan of `\n` before `try` and after `}`. :smile:

File: syncany-lib/src/main/java/org/syncany/operations/init/ApplicationLink.java [hunk 15]
```diff
@@ -361,14 +364,19 @@ private TransferSettings createTransferSettings(byte[] plaintextPluginSettingsBy
...
-	private byte[] getPlaintextStorageXml() throws Exception {
+	private byte[] getPlaintextStorageXml() throws IOException, SerializableException {
 		ByteArrayOutputStream plaintextByteArrayOutputStream = new ByteArrayOutputStream();
 		DataOutputStream plaintextOutputStream = new DataOutputStream(plaintextByteArrayOutputStream);
 		plaintextOutputStream.writeInt(transferSettings.getType().getBytes().length);
 		plaintextOutputStream.write(transferSettings.getType().getBytes());
 
 		GZIPOutputStream plaintextGzipOutputStream = new GZIPOutputStream(plaintextOutputStream);
-		new Persister(new Format(0)).write(transferSettings, plaintextGzipOutputStream);
+		try {
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/init/ApplicationLink.java]: \n before try and after }

File: syncany-lib/src/main/java/org/syncany/operations/init/ConnectOperation.java [hunk 16]
```diff
@@ -83,119 +86,123 @@ public ConnectOperation(ConnectOperationOptions options, UserInteractionListener
...
 			}
-		}
 
-		result.setResultCode(ConnectResultCode.OK);
+			result.setResultCode(ConnectResultCode.OK);
+		}
+		catch (StorageException | IOException | ConfigException e) {
+			throw new OperationException(e);
+		}
 		return result;
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/init/ConnectOperation.java]: }\n -> }\n\n

File: syncany-lib/src/main/java/org/syncany/operations/plugin/PluginException.java [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.syncany.operations.plugin;
+
+
+/**
+ * Base class for exceptions related to installed/uninstallable plugins.
+ * 
+ * @author Stefan Hugtenburg
+ *
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/plugin/PluginException.java]: - Remove unnecessary new lines
- Move constant to top

File: syncany-lib/src/main/java/org/syncany/operations/plugin/PluginNotInstalledException.java [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.syncany.operations.plugin;
+
+/**
+ * Exception that can be used when a plugin is not installed yet needs/wants to be used.
+ * 
+ * @author Stefan Hugtenburg
+ *
+ */
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/plugin/PluginNotInstalledException.java]: - Remove \n's 
- Move constant to top

File: syncany-lib/src/main/java/org/syncany/operations/plugin/PluginOperation.java [hunk 19]
```diff
@@ -112,28 +118,50 @@ public PluginOperation(Config config, PluginOperationOptions options) {
...
 	@Override
-	public PluginOperationResult execute() throws Exception {
+	public PluginOperationResult execute() throws OperationException {
 		result.setAction(options.getAction());
 
 		switch (options.getAction()) {
-			case LIST:
+		case LIST:
+			try {
 				return executeList();
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/plugin/PluginOperation.java]: Same as DaemonOp: Awkward construction with the try/catch+switch/case

File: syncany-lib/src/main/java/org/syncany/operations/plugin/UninstallablePluginException.java [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.syncany.operations.plugin;
+
+
+/**
+ * Class that holds information about plugins that can not be installed.
+ * 
+ * @author Stefan Hugtenburg
+ *
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/plugin/UninstallablePluginException.java]: \n

File: syncany-lib/src/main/java/org/syncany/operations/up/BlockingTransfersException.java [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.syncany.operations.up;
+
+/**
+ * @author Jesse Donkervliet
+ *
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/up/BlockingTransfersException.java]: Remove unnecessary \n

File: syncany-lib/src/main/java/org/syncany/operations/up/UpOperation.java [hunk 22]
```diff
@@ -106,145 +119,301 @@ public UpOperation(Config config, UpOperationOptions options) {
...
 
 	@Override
-	public UpOperationResult execute() throws Exception {
+	public UpOperationResult execute() throws OperationException {
 		logger.log(Level.INFO, "");
 		logger.log(Level.INFO, "Running 'Sync up' at client " + config.getMachineName() + " ...");
 		logger.log(Level.INFO, "--------------------------------------------");
 
 		fireStartEvent();
+		try {
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/up/UpOperation.java]: - \n before 'try'
- This function is now becoming rather ugly. All the try/catches don't make it better. I would love to see some refactoring on that (mainly splitting into smaller functions) -- again, not in this pull though

File: syncany-lib/src/main/java/org/syncany/operations/up/UpOperation.java [hunk 23]
```diff
@@ -106,145 +119,301 @@ public UpOperation(Config config, UpOperationOptions options) {
...
-		boolean resuming = false;
+			BlockingQueue<DatabaseVersion> databaseVersionQueue = new LinkedBlockingQueue<>();
+			boolean resuming = false;
 
-		if (options.isResume()) {
-			// If we want to resume, we look for the local files that contain data about a resumable transaction.
-			newDatabaseVersion = attemptResumeTransaction();
+			if (options.isResume()) {
+				Collection<Long> versionsToResume = transferManager.loadPendingTransactionList();
+				if (versionsToResume != null && versionsToResume.size() != 0) {
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/up/UpOperation.java]: \n before if

File: syncany-lib/src/main/java/org/syncany/operations/up/UpOperation.java [hunk 24]
```diff
@@ -106,145 +119,301 @@ public UpOperation(Config config, UpOperationOptions options) {
...
-			if (newDatabaseVersion != null) {
-				logger.log(Level.INFO, "Found local transaction to resume.");
-				resuming = true;
+					try {
+						remoteTransactionsToResume = attemptResumeTransactions(versionsToResume);
+					}
+					catch (DeserializableException e1) {
+						throw new StorageException(e1);
+					}
+					Collection<DatabaseVersion> remoteDatabaseVersionsToResume = attemptResumeDatabaseVersions(versionsToResume);
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/up/UpOperation.java]: \n after }

File: syncany-lib/src/main/java/org/syncany/operations/up/UpOperation.java [hunk 25]
```diff
@@ -106,145 +119,301 @@ public UpOperation(Config config, UpOperationOptions options) {
...
+	 * @throws IOException 
+	 */
+	private int executeTransactions(BlockingQueue<DatabaseVersion> databaseVersionQueue) throws IOException, StorageException {
+		return executeTransactions(databaseVersionQueue, null, null);
+	}
 
-			ChangeSet localChanges = result.getStatusResult().getChangeSet();
-			List<File> locallyUpdatedFiles = extractLocallyUpdatedFiles(localChanges);
+	/**
+	 *	Transfers the given {@link DatabaseVersion} objects to the remote.
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/up/UpOperation.java]: Malformatted comment " " instead of "\t"?

File: syncany-lib/src/main/java/org/syncany/operations/up/UpOperation.java [hunk 26]
```diff
@@ -106,145 +119,301 @@ public UpOperation(Config config, UpOperationOptions options) {
...
+	 *	@param databaseVersionQueue The {@link DatabaseVersion} objects to send to the remote.
+	 *	@param remoteTransactionsToResume {@link RemoteTransaction} objects that correspond to the given {@link DatabaseVersion} objects.
+	 *	@param transactionRemoteFileToResume The file on the remote that was used for the specific transaction that was interrupted.
+	 * @throws IOException 
+	 * @throws StorageException 
+	 * @throws  
+	 */
+	private int executeTransactions(BlockingQueue<DatabaseVersion> databaseVersionQueue, Iterator<RemoteTransaction> remoteTransactionsToResume,
+			TransactionRemoteFile transactionRemoteFileToResume) throws IOException, StorageException
+	{
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/up/UpOperation.java]: { in previous line

File: syncany-lib/src/main/java/org/syncany/operations/up/UpOperationOptions.java [hunk 27]
```diff
@@ -24,6 +24,12 @@
 
 @Root(name = "up")
 public class UpOperationOptions implements OperationOptions {
+	// The transaction size limit determines how much data (in bytes) is combined into a single transaction during an
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/up/UpOperationOptions.java]: /* */ instead of //

File: syncany-lib/src/main/java/org/syncany/operations/up/UpOperation.java [hunk 28]
```diff
@@ -641,19 +802,46 @@ private DatabaseVersion attemptResumeTransaction() throws Exception {
...
 
-			// Writing database representation of new database version to state dir
-			MemoryDatabase memoryDatabase = new MemoryDatabase();
-			memoryDatabase.addDatabaseVersion(newDatabaseVersion);
+			// Write the list of version number to a file, before serializing any transactions!
+			// This ensures that no transaction files can exist without a "reference" to them.
+			File transactionListFile = config.getTransactionListFile();
+			PrintWriter transactionListWriter = new PrintWriter(new OutputStreamWriter(
+					new FileOutputStream(transactionListFile), "UTF-8"));
+			for (Long databaseVersion : databaseVersionClocks) {
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/up/UpOperation.java]: \n before 'for' and after '}'

File: syncany-lib/src/main/java/org/syncany/plugins/transfer/to/TransactionTO.java [hunk 29]
```diff
@@ -83,10 +84,16 @@ public static TransactionTO load(Transformer transformer, File transactionFile)
...
+			return new Persister().read(TransactionTO.class, is);
+		}
+		catch (Exception e) {
+			throw new DeserializableException(e);
+		}
 	}
 
-	public void save(Transformer transformer, File transactionFile) throws Exception {
+	public void save(Transformer transformer, File transactionFile) throws SerializableException, IOException
+	{
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/plugins/transfer/to/TransactionTO.java]: { in previous line

**Comment by binwiederhier** [comment]
@MrHug 
Thanks very much for this awesome refactoring. All of the changes make total sense to me and apart from my OCD-type affection to new lines, everything looks great. 

As you said, this depends on #437, so we'll have to wait for that to be merged until we merge this. As we move along, it's important to keep this branch updated to avoid a big mess when we finally merge it.

Again, awesome work. Thank you!

**Comment by MrHug** [comment]
I'll try to find some time to do something about these comments asap :wink:

Op wo 6 mei 2015 om 01:47 schreef Philipp C. Heckel <
notifications@github.com>:

File: syncany-lib/src/main/java/org/syncany/operations/ActionFileHandler.java (hunk 1 above)
Comment (MrHug) [inline syncany-lib/src/main/java/org/syncany/operations/ActionFileHandler.java]: Oops, I thought I'd already put that back (I'm not sure where this came from. Fixed in new commit).

File: syncany-lib/src/main/java/org/syncany/operations/daemon/messages/api/XmlMessageFactory.java (hunk 7 above)
Comment (MrHug) [inline syncany-lib/src/main/java/org/syncany/operations/daemon/messages/api/XmlMessageFactory.java]: Hmm I missed the `SerializationException` I guess that one would be alright. The reason I have two different ones is that they distinguish between reading and writing the serialization. It gives a tad bit more detail on what went wrong (incorrect/corrupted data that can not be converted to a Java Object, or vice versa). I haven't touched this yet, so let me know what you would prefer.

If you'd prefer only one exception, then the `SerializationException` indeed looks like a good one to use.

File: syncany-lib/src/main/java/org/syncany/operations/down/DownOperation.java (hunk 10 above)
Comment (MrHug) [inline syncany-lib/src/main/java/org/syncany/operations/down/DownOperation.java]: Yeah I took a look, but that should indeed be another PR as that would probably be very significant changes with way more chance of me breaking something ;)

File: syncany-lib/src/main/java/org/syncany/operations/down/actions/ChangeFileSystemAction.java (hunk 12 above)
Comment (MrHug) [inline syncany-lib/src/main/java/org/syncany/operations/down/actions/ChangeFileSystemAction.java]: Jup, so saving it causes this to be messed up again (I have format on save enabled). Would you like me to edit this in vim or fix it in some other way?

File: syncany-lib/src/main/java/org/syncany/operations/down/actions/FileCreatingFileSystemAction.java (hunk 14 above)
Comment (MrHug) [inline syncany-lib/src/main/java/org/syncany/operations/down/actions/FileCreatingFileSystemAction.java]: I've put the line in there, but normally I would not advocate this as the Exceptions we want to catch are from the `assembleFileToCatch` call and not the other one :wink:

File: syncany-lib/src/main/java/org/syncany/operations/up/BlockingTransfersException.java (hunk 21 above)
Comment (MrHug) [inline syncany-lib/src/main/java/org/syncany/operations/up/BlockingTransfersException.java]: I blame @jdonkervliet for this one, but will fix it regardless ;)

File: syncany-lib/src/main/java/org/syncany/operations/up/UpOperation.java (hunk 22 above)
Comment (MrHug) [inline syncany-lib/src/main/java/org/syncany/operations/up/UpOperation.java]: Jup probably in the same pull as the Down op :wink:

**Comment by MrHug** [comment]
Did some work, though now there seem to be merge conflicts :( (not sure if they are due to my changes or not)

I also commented on some of your comments above, but it seems they are still removed if the code is changed...

File: syncany-lib/src/main/java/org/syncany/operations/daemon/messages/api/XmlMessageFactory.java (hunk 7 above)
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/daemon/messages/api/XmlMessageFactory.java]: Can't we usually determine from the stack trace if something went wrong during serialization or deserialization? If not, I'm all for different exceptions. If not, then it doesn't serve any purpose.

File: syncany-lib/src/main/java/org/syncany/operations/down/DownOperation.java (hunk 10 above)
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/down/DownOperation.java]: All we do is throw a different exception, and we're either catching or ignore-ing that exception anyway.

File: syncany-lib/src/main/java/org/syncany/operations/down/actions/ChangeFileSystemAction.java (hunk 12 above)
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/down/actions/ChangeFileSystemAction.java]: naaahh... all good. I really need to enable that as well...

File: syncany-lib/src/main/java/org/syncany/chunk/InvalidSettingsException.java [hunk 30]
```diff
@@ -0,0 +1,37 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.syncany.chunk;
+
+/**
+ * Thrown by Transformer init methods when their settings are not valid.
+ * 
+ * @author Stefan Hugtenburg
+ *
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/chunk/InvalidSettingsException.java]: -\n

File: syncany-lib/src/main/java/org/syncany/chunk/InvalidSettingsException.java [hunk 31]
```diff
@@ -0,0 +1,37 @@
...
+ */
+public class InvalidSettingsException extends Exception {
+
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/chunk/InvalidSettingsException.java]: -\n

File: syncany-lib/src/main/java/org/syncany/chunk/InvalidSettingsException.java [hunk 32]
```diff
@@ -0,0 +1,37 @@
...
+	/**
+	 * @param msg Message for the Exception.
+	 */
+	public InvalidSettingsException(String msg) {
+		super(msg);
+	}
+
+	private static final long serialVersionUID = -280945343359999227L;
+
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/chunk/InvalidSettingsException.java]: - -\n
- Move constant to top

**Comment by binwiederhier** [comment]
I just noticed that you haven't merged the latest develop. Can you do that before we continue the review? I stumbled across `FILE_TRANSACTION_PATTERN` and thought that was part of Pim's commit. It shouldn't be in this diff, right?

File: syncany-lib/src/main/java/org/syncany/operations/daemon/messages/api/XmlMessageFactory.java (hunk 7 above)
Comment (MrHug) [inline syncany-lib/src/main/java/org/syncany/operations/daemon/messages/api/XmlMessageFactory.java]: Sure, but know you'd know by just looking at the name of the Exception... I'm perfectly fine with SerializationException as well, but don't have the time to go over all of the changes now/this weekend.

**Comment by MrHug** [comment]
@binwiederhier Processed the last bit of feedback and merged develop into it. Let me know if anything else needs doing, though it might be a while before I get to it (little time and all that :wink: ).

**Comment by binwiederhier** [comment]
Besides my usual style complains, I have only one concern: This introduces a lot of try/catch blocks that do not make the code easier to read. While I understand the "don't use Exception" argument, I don't necessrily agree with it because of the reduced readability.

However, the excessive try/catching can be reduced by some refactoring; which will happen anyway. 

Please don't merge this just yet. I would like to take care of some OCD-type style issues myself, if nobody minds :)

**Comment by MrHug** [comment]
The main idea of this PR was to show that you've been using `throws Exception` way too often. This refactoring into different exceptions should indeed be the first step of a code cleanup that happens on a larger scale than just these "minor" revisions. Unfortunately I do not think I have time any time soon to do anything for this, but if I do I'll let you know and will probably start "real refactoring" :wink:

**Comment by binwiederhier** [comment]
Superseded by #498.
